sklearn/utils/sparsefuncs_fast.pyx

```diff
@@ -48,6 +50,63 @@ def csr_row_norms(X):
 @cython.boundscheck(False)
 @cython.wraparound(False)
 @cython.cdivision(True)
+def sparse_mean_variance(X):
+    """Compute the variance of a sparse matrix.
+    Will copy data to an array of doubles if input data
+    are not np.float32 or np.float64 types.
+
+    Parameters
+    ----------
+    X: Scipy sparse matrix, shape (n_samples, n_features)
+        Input data.
+
+    Returns
+    -------
+
+    mean: float scalar
+        Mean of array
+
+    variance: float scalar
+        Variance of array
+    """
+    cdef:
+        double sum = 0
+        double sum_squares = 0
+        double variance = 0
+        double n_elements = X.shape[0] * X.shape[1]
+        np.ndarray[DOUBLE, ndim=1, mode="c"] X_data_d
+        np.ndarray[FLOAT, ndim=1, mode="c"] X_data_f
+        int len_data
+        double datum
+    len_data = len(X.data)
+
+    X_data = X.data
+    if X_data.dtype not in (np.float64, np.float32):
+        warnings.warn("Copying sparse array data to array of "
+                      "doubles for computation.")
+        X_data = np.asarray(X_data, dtype=np.float64)  # Copy!
+    if X_data.dtype == np.float64:
+        X_data_d = X_data
+        for i_datum in xrange(len_data):
+            datum = X_data_d[i_datum]
+            sum += datum
+            sum_squares += datum * datum
+    elif X_data.dtype == np.float32:
+        X_data_f = X_data
+        for i_datum in xrange(len_data):
+            datum = X_data_f[i_datum]
+            sum += datum
+            sum_squares += datum * datum
```

Comment:
The `float16` is not supported in C, so there's no way to include it in compile Cython code. The best reference I found Googling on that is at https://groups.google.com/forum/#!topic/cython-users/a_gRQJzauJ4 . That thread is old now, but my current version of Cython also does not define a float16 type.

`X.data.mean()` and `X.data.var()` aren't the same as `X.mean()` and `X.var()`, since they ignore the implicit zeros in the sparse array. The scipy sparse matrices support `X.mean()`, but for some reason not `X.var()`. I can think of two ways to do this from pure Python:
`variance = (arr.data ** 2).sum() / (arr.shape[0] * arr.shape[1]) - arr.mean() ** 2`
or 
```
def meanvar_sparse(arr):
    data_mean = arr.data.mean()
    data_var = arr.data.var()
    ndata = len(arr.data)
    narr = arr.shape[0] * arr.shape[1]
    data_sum = data_mean * ndata
    data_sum_squares = (data_var + (data_mean * data_mean)) * ndata
    
    arr_mean = (data_sum / narr)
    arr_var = data_sum_squares / narr - arr_mean ** 2
    
    return arr_mean, var
```

Both are slower than the Cython code, and the first uses extra memory by creating a new array of length `len(X.data)` as an intermediate result. Since I'd need a new function anyway, and the scikit-learn code already included Cython functions to compute means and variances along columns and rows of sparse matrices, it seemed reasonable to add another Cython function for the entire array variance. 

I actually tried fused types before I submitted this PR. It does make the code simpler, but I was surprised to find that it was slower. Since there's not a large number of cases to support (float32 and float64 seem like the only ones that matter), I opted for the extra explicit case rather than fused types.

Following are some results of speed testing with `%timeit`. I'm comparing the current state of this PR (calling `sparsefuncs.mean_variance_axis(arr, axis=None)`) to a Cython function `_floating_mean_variance` defined for fused float32 and float64 types (fully typed, so that there's no Python lines left in the annotate), to the Python `meanvar_sparse` defined above. My test data are randomly-generated `csr_matrix` objects with either shape 5kx5k and 100k non-zero entries, or 50k x 50k and 1M or 10M non-zero entries. Results:

Non-zero entries | No fused types | Fused types | Python
-----------------------|----------------------|------------------|----------
100,000 | 112 us | 177 us | 325 us
1,000,000 | 1.21 ms | 1.36 ms | 5.25 ms
10,000,000 | 11.6 ms | 11.8 ms | 73 ms

If I switch the `_floating_mean_variance` function to use np.float64_t instead of a fused type, its speed equals the current `sparse_mean_variance`, which confirms that the extra time is overhead in detecting the type and choosing which function to use. The constant overhead is not large, but it still didn't seem worth using the fused types here.

Given this, my preference is to remain with the current `sparse_mean_variance` function.